### DISCUSSION TITLE
Recommend approach to test derived atoms

### DISCUSSION THREAD

**Discussion by soruban** [pmndrs/jotai#1613] [Q&A]
Hi, I have been playing around jotai and was wondering what is the best way to test a heavily derived atom e.g.

```typescript
// Assume that foo & bar are atoms that rely on server data or some complex operations too.
const derivedAtom = atom((get) => get(foo) + get(bar) + .... 

const MyComponent = () => {
  const value = useAtomValue(derivedAtom);

  return <div>{value}</div>
}

describe('test1', () => {
  const MyComponentProvider = () => {
    return (
      <MyComponent initialValues={[[derivedAtom, 100]]}>
        <Foo />
      </MyComponent>
    );
  };

  it('should display the value', () => {
    render(<MyComponent />);

    expect(screen.queryByText(100)).not.toBeNull();
  });
});
```

The above is great, this component only uses the derivedAtom and I can easily initialize it in the Provider, however this warning is triggered:
`Found initial value for derived atom which can cause unexpected behavior { toString: [Function: toString], read: [Function (anonymous)] }`

It makes sense then... is there a better warning-free recommended approach? 
Or as long as one fully understand the consequences the above test is the right way?

**Comment by dai-shi** [comment]
I would probably mock the entire `derivedAtom`.

**Reply by soruban** [reply]
Do you mean mock every atom that derivedAtom is made of?

**Reply by dai-shi** [reply]
Yeah, make it like `export const derivedAtom = atom(() => 100)` using `__mocks__` with [jest](https://jestjs.io/docs/manual-mocks).

**Reply by soruban** [reply]
Indeed, that is an approach that would work but the developer experience would suffer as the project grows.

Let me use a more concrete example, assume an application with user configurable widgets:
```typescript
// In main-atoms.ts
const config = atom(fetchConfig);
const flag = atom(get => get(config).flag)
// ...and more


// In user-atoms.ts
const userSettings = atom(userSettings);
const userName = atom(get => `${get(userSettings).name} ${get(userSettings).lastName}`;
// ...and more

// In component-atoms.ts
const componentConfig = atom((get) => doSomeTransform(get(config), get(userSettings));
const componentData = atom((get) => fetchData(get(userSettings).foo));
const componentSettings = atom((get) => fetchData(get(userSettings).bar));
// ...and more

// Component.ts
const MyComponent = () => {
  const flag = atom(flag);
  const name = useAtomValue(userName);
  const compSettings = useAtomValue(componentSettings);

  if(flag) {
    return <div>{/* Some alternate view */}</div>
  } else {
    return <div><div>{compSettings.foo}</div>{name}</div>
  }
}
```

If we wanted to test MyComponent, the __mocks__ approach developer experience has the following issues:
- we'd need to create mocks for each file.
- every time an atom is added/deleted/updated, the mock file will need to be maintained.
- if we wanted to mock some derived atoms and not others, it would be difficult.

Overall, I think the current solution which triggers a warning is the most dev friendly & flexible, is there a way to create a "TestProvider" which will be used for UTs only & allow derived atoms overrides (and possibly other testing features)?

**Reply by dai-shi** [reply]
> is there a way to create a "TestProvider" which will be used for UTs only & allow derived atoms overrides (and possibly other testing features)?

That might be possible, and I especially like such "other testing features".

In Jotai v2 API, we no longer have `initialValues` 🤯 .
I think it makes sense to provide `createTestStore`.
I would very much like to get your idea as I'm not super experienced with testing.
You may need to learn the v2 API first.
Link to links: https://twitter.com/dai_shi/status/1601092344262316033

Meanwhile, my suggestion is to hide the warning by mocking `console.warn`.

**Reply by soruban** [reply]
I see, let me take a look at the new API and get back to you, happy to connect to discuss testing approaches further too!

**Reply by patricktruong-cb** [reply]
Any update on this issue I am also having trouble mocking this for tests

**Reply by dvd900** [reply]
In my application and in your example, i've noticed that for the most part only a few of the atoms are actually consumed while the rest represent logic on how to derive them. For ease of testing, I often alias out the useAtom.

as an example: 
```typescript
// In component-atoms.ts
const componentSettings = atom((get) => fetchData(get(userSettings).bar));
export const useComponentSettings = () => useAtomValue(componentSettings);

// Component.ts
const MyComponent = () => {
...
  const compSettings = useComponentSettings();
}

``` 

This way I can easily mock the `useComponentSettings` function using jest, without having to go mocking my way all the way up the dependency list. I can write other tests that unit test the actual logic in the atom, and test my component in a more pure way. Far from a perfect option, but has proven useful for me.
